src/plugins/api/index_pattern_endpoints.js

```diff
@@ -0,0 +1,221 @@
+const getMappings = require('./lib/get_mappings');
+const stitchPatternAndMappings = require('./lib/stitch_pattern_and_mappings');
+const {templateToPattern, patternToTemplate} = require('./lib/convert_pattern_and_template_name');
+const removeDeprecatedFieldProps = require('./lib/remove_deprecated_field_props');
+
+export default function (server) {
+
+  let handleESError = function (error) {
+    if (error instanceof esErrors.ConnectionFault ||
+      error instanceof esErrors.ServiceUnavailable ||
+      error instanceof esErrors.NoConnections ||
+      error instanceof esErrors.RequestTimeout) {
+      return Boom.serverTimeout();
+    } else if (error instanceof esErrors.Conflict) {
+      return Boom.conflict();
+    } else if (error instanceof esErrors[403]) {
+      return Boom.forbidden();
+    } else if (error instanceof esErrors.NotFound) {
+      return Boom.notFound();
+    } else if (error instanceof esErrors.BadRequest || error instanceof TypeError) {
+      return Boom.badRequest(error);
+    } else {
+      return error;
+    }
+  };
+
+  server.route({
+    path: '/api/index-patterns',
+    method: 'GET',
+    handler: function (req, reply) {
+      let client = server.plugins.elasticsearch.client;
+
+      client.search({
+        index: '.kibana',
+        type: 'index-pattern',
+        body: {
+          query: {
+            match_all: {}
+          }
+        }
+      }).then(function parseResults(results) {
+        const hits = results.hits.hits;
+        return _.map(hits, (patternHit) => {
+          patternHit._source.fields = JSON.parse(patternHit._source.fields);
+          return patternHit._source;
+        });
+      }).then((patterns) => {
+        return Promise.map(patterns, (pattern) => {
+          return getMappings(pattern.title, client).catch(() => {
+            return {};
+          }).then((mappings) => {
+            return stitchPatternAndMappings(pattern, mappings);
+          });
+        });
+      }).then(removeDeprecatedFieldProps)
+        .then((patterns) => {
+          reply(patterns);
+        }, function (error) {
+          reply(handleESError(error));
+        });
+    }
+  });
+
+  server.route({
+    path: '/api/index-patterns/{id}',
+    method: 'GET',
+    handler: function (req, reply) {
+      let client = server.plugins.elasticsearch.client;
+      let pattern = req.params.id;
+
+      Promise.join(
+        client.get({
+          index: '.kibana',
+          type: 'index-pattern',
+          id: req.params.id
+        }).then((result) => {
+          result._source.fields = JSON.parse(result._source.fields);
+          return result._source;
+        }),
+        getMappings(pattern, client),
+        stitchPatternAndMappings
+      ).then(removeDeprecatedFieldProps)
+        .then(function (pattern) {
+          reply(pattern);
+        }, function (error) {
+          reply(handleESError(error));
+        });
+    }
+  });
+
+  server.route({
+    path: '/api/index-patterns',
```

Comment:
I'm really not a fan of camelCase in urls. I think hyphens and underscores are much more prevalent in api design and since this is an interface we're exposing to the outside world, I don't think our internal naming conventions should dictate how we form these urls.